src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.cs

```diff
@@ -75,7 +75,8 @@ private void OnDiagnosticsUpdated(object? sender, DiagnosticsUpdatedArgs e)
                 return;
             }
 
-            var document = e.Solution.GetDocument(e.DocumentId);
+            var document = e.Solution.GetDocument(e.DocumentId)
+                ?? e.Solution.GetProject(e.DocumentId.ProjectId)?.TryGetSourceGeneratedDocumentForAlreadyGeneratedId(e.DocumentId);
 
             // Open documents *should* always have their SourceText available, but we cannot guarantee
             // (i.e. assert) that they do.  That's because we're not on the UI thread here, so there's
```

Comment:
How is FindCorrespondingEditorTextSnapshot below working in the source generator case -- there's an icky problem here where we may have new diagnostics but the buffer might not be in sync yet?

Reply:
➡️ I can't find any purpose for this code.

Reply:
@CyrusNajmabadi You remember more here? I see I reviewed that PR but not sure I remember the full details here.

Reply:
yeah, ican see why we do this... but i'm also not sure what the best solution forward is.  for one thing, i'm not sure why this is a mapping from diag to snapshot, instead of diag to Document (since that's waht we have now). then, later, when actually producing tags for another snapshot (which is async) we can then figure out what snapshot this doc actually applied to, and then map the diags forward.

Reply:
Offline discussion concluded here that we're doing this because when we produce diagnostics, that's running on a potentially older snapshot of an editor. We do this to capture the snapshot that maps to those old diagnostic, so we can then map them forward a-la tracking spans to the current snapshot. In the case of generated files, we have the reverse problem: we may have generated diagnostics from the future that need to be brought back. Simply not updating this code will do that (along with some clamping to the file).